disperser/dataapi/blobs_handlers.go

```diff
@@ -96,3 +113,111 @@ func convertMetadataToBlobMetadataResponse(metadata *disperser.BlobMetadata) (*B
 		BlobStatus:              metadata.BlobStatus,
 	}, nil
 }
+
+func (s *server) getBlobMetadataByBatchesWithLimit(ctx context.Context, limit int) ([]*Batch, []*disperser.BlobMetadata, error) {
+	var (
+		blobMetadatas   = make([]*disperser.BlobMetadata, 0)
+		batches         = make([]*Batch, 0)
+		blobKeyPresence = make(map[string]struct{})
+		batchPresence   = make(map[string]struct{})
+	)
+
+	for skip := 0; len(blobMetadatas) < limit && skip < limit; skip += maxQueryBatchesLimit {
+		batchesWithLimit, err := s.subgraphClient.QueryBatchesWithLimit(ctx, maxQueryBatchesLimit, skip)
+		if err != nil {
+			s.logger.Error("Failed to query batches", "error", err)
+			return nil, nil, err
+		}
+
+		if len(batchesWithLimit) == 0 {
+			break
+		}
+
+		for i := range batchesWithLimit {
+			s.logger.Debug("Getting blob metadata", "batchHeaderHash", batchesWithLimit[i].BatchHeaderHash)
+			var (
+				batch = batchesWithLimit[i]
+			)
+			if batch == nil {
+				continue
+			}
+			batchHeaderHash, err := ConvertHexadecimalToBytes(batch.BatchHeaderHash)
+			if err != nil {
+				s.logger.Error("Failed to convert batch header hash to hex string", "error", err)
+				continue
+			}
+			batchKey := string(batchHeaderHash[:])
+			if _, found := batchPresence[batchKey]; !found {
+				batchPresence[batchKey] = struct{}{}
+			} else {
+				// The batch has processed, skip it.
+				s.logger.Error("Getting duplicate batch from the graph", "batch header hash", batchKey)
+				continue
+			}
+
+			metadatas, err := s.blobstore.GetAllBlobMetadataByBatch(ctx, batchHeaderHash)
+			if err != nil {
+				s.logger.Error("Failed to get blob metadata", "error", err)
+				continue
+			}
+			for _, bm := range metadatas {
+				blobKey := bm.GetBlobKey().String()
+				if _, found := blobKeyPresence[blobKey]; !found {
+					blobKeyPresence[blobKey] = struct{}{}
+					blobMetadatas = append(blobMetadatas, bm)
+				} else {
+					s.logger.Error("Getting duplicate blob key from the blobstore", "blobkey", blobKey)
+				}
+			}
+			batches = append(batches, batch)
+			if len(blobMetadatas) >= limit {
+				break
+			}
+		}
+	}
+
+	if len(blobMetadatas) >= limit {
+		blobMetadatas = blobMetadatas[:limit]
+	}
+
+	return batches, blobMetadatas, nil
+}
+
+func (s *server) getBlobMetadataByBatchHeaderHashWithLimit(ctx context.Context, batchHeaderHash string, limit int32, exclusiveStartKey *disperser.BatchIndexExclusiveStartKey) ([]*disperser.BlobMetadata, *disperser.BatchIndexExclusiveStartKey, error) {
+	batchHeaderHashBytes, err := ConvertHexadecimalToBytes([]byte(batchHeaderHash))
+	if err != nil {
+		s.logger.Error("Failed to convert batch header hash to hex string", "error", err)
+		return nil, nil, err
+	}
+
+	var allMetadata []*disperser.BlobMetadata
+	var nextKey *disperser.BatchIndexExclusiveStartKey = exclusiveStartKey
+
+	const maxLimit int32 = 1000
```

Comment:
A batch can have more than 1000 blobs easily (just 2 blobs/s)

Reply:
Yeah i've seen a batch have up to 19000 blobs. I'm setting some reasonable max limit on the page size. Do you think 10,000 is a better upper bound?

Reply:
Trying with 10,000 takes around 13 seconds with a 27 megabyte response. While 1000 takes less than 2 seconds with 2 megabyte response. 

I think it's better with smaller pages since it can also be parallelized.